lms/djangoapps/lti_provider/users.py

```diff
@@ -28,14 +28,34 @@ def authenticate_lti_user(request, lti_user_id, lti_consumer):
     If the currently logged-in user does not match the user specified by the LTI
     launch, log out the old user and log in the LTI identity.
     """
+    lis_email = request.POST.get("lis_person_contact_email_primary")
+
     try:
         lti_user = LtiUser.objects.get(
             lti_user_id=lti_user_id,
             lti_consumer=lti_consumer
         )
-    except LtiUser.DoesNotExist:
+    except LtiUser.DoesNotExist as exc:
         # This is the first time that the user has been here. Create an account.
-        lti_user = create_lti_user(lti_user_id, lti_consumer)
+        if lti_consumer.require_user_account:
+            # Verify that the email from the LTI Launch and the logged-in user are the same
+            # before linking the LtiUser with the edx_user.
+            if request.user.is_authenticated and request.user.email == lis_email:
+                lti_user = create_lti_user(lti_user_id, lti_consumer, lis_email)
+            else:
+                # Ask the user to login before linking.
+                raise PermissionDenied() from exc
+        else:
+            lti_user = create_lti_user(lti_user_id, lti_consumer)
+
+    # If auto-linking is enabled, the lti_user should be linked to the logged-in user
+    if (
+        lti_consumer.require_user_account and
+        request.user.is_authenticated and
+        lti_user.edx_user != request.user
+    ):
+        lti_user.edx_user = request.user
+        lti_user.save()
```

Comment:
@tecoholic, let's assume we have Alice and Bob who are the learners. Is the following scenario correct?
1. Alice has already used LTI, so her Canvas and Open edX accounts are linked.
2. Bob used a shared computer in a classroom and left his Open edX account logged in.
3. Alice uses the same shared computer, logs into Canvas, and goes to a Canvas page with the LTI content from Open edX.
4. We have one of the following outcomes:
    1.  If Bob had used LTI before, this raises an `IntegrityError` (due to the `'lti_consumer', 'lti_user_id'` uniqueness violation).
    2. Otherwise, Bob's account is accidentally "hijacked" by Alice.
        1. Bob's account will get re-linked again (to the correct account) the next time he uses the LTI, but it is still a security concern.

In this scenario, we want to switch the Open edX user to the one linked to the `LtiUser` (with the `switch_user` function).
The only real-world usage I can think of for re-linking the accounts is moving the existing (automatically created) LTI users to "real" Open edX accounts after enabling an existing provider's `require_user_account`  option. If this is our goal, we should verify if the `User`'s email ends with `settings.LTI_USER_EMAIL_DOMAIN`. However, all existing progress of such users will be lost this way, so we should be sure that we want to do this (instead of, e.g., changing the emails of their automatically created accounts).

Reply:
@Agrendalath First up, I am just astounded at how many of the issues I miss and you catch. And I entirely understand that's the function of the work-review cycle and really appreciate it.

And yes, the only real reason to switch the user here is to make sure we don't have a mix of real users and anonymous users on a client if/when the flag is turned on. I had my reservations about it, but this is something I totally missed

> However, all existing progress of such users will be lost this way, so we should be sure that we want to do this 

So, I think it is best to remove the re-linking part now. I will update the PR.